spectre_analyses/tasks.py

```diff
@@ -84,23 +80,20 @@ def __exit__(self, *args):
         signal.signal(_CELERY_REVOKE, self._previous_handler)
 
 
-class cleanup(hijack_revoke):
-    def __init__(self, path: str):
-        super(hijack_revoke, self).__init__()
-        self.path = path
-
-    def __call__(self, signal_number, stack_frame):
-        if os.path.exists(self.path):
-            shutil.rmtree(self.path, ignore_errors=True)
+def cleanup(path: str, *_):
+    """Clean up analysis directory"""
+    if os.path.exists(path):
+        shutil.rmtree(path, ignore_errors=True)
 
 
 @contextmanager
 def _open_analysis(dataset_name: str, algorithm_name: str, analysis_name: str):
     path_components = dataset_name, algorithm_name, analysis_name
     analysis_root = os.path.join(STATUS_PATHS['processing'], *path_components)
     os.makedirs(analysis_root)
+    cleanup_root = partial(cleanup, analysis_root)
     try:
-        with cleanup(analysis_root):
+        with hijack_revoke(cleanup_root):
             yield analysis_root
```

Comment:
I’m not really convinced, that `hijack_revoke` is a good name; it hardly describes what’s happening. This is basically a _signal trap_ for `SIGTERM`. How about refactoring it in the lines of:
```python
with trap(signal.SIGTERM, cleanup_root):
    yield ...
```
This makes it explicit that `cleanup_root` should fire for `SIGTERM` during execution of the `with` block.

Reply:
I treat specific signal as an implementation detail of Celery, which `SIGTERM` is in this case. Few more edits, and we will all wonder, why are we trapping the signal at all. `hijack_revoke` tells you, that you are intercepting process of task revoking, which is exactly, what happens here.

We do not need generic signal trap. YAGNI, as wise man @mg6 tells me each time. :)

Reply:
This is not an implementation detail; it is the standard way to stop programs under POSIX, a kind of protocol that all Unix programs obey.

No matter how you stop the program: with `kill` (or `docker kill`), by running `docker stop` or by canceling task through Celery, etc. — you always end up sending `SIGTERM` as a request to *gracefully terminate* the process. Only after the program fails to gracefully terminate in predefined time, `SIGKILL` should be used to kill the program without any chance of cleanup (you cannot trap `SIGKILL`). 

Therefore, I stand by my initial point of view: `hijack_revoke` is a totally misleading name; what you actually do is a typical *signal trap*, with signal being SIGTERM. You could even refactor it in such a way, that it is the default signal, for brevity:

```python
def __init__(self, handler, signal=signal.SIGTERM):
    self._previous_handler = signal.SIG_DFL
    self._handler = handler
    self._signal = signal
```

```python
with trap(cleanup_root):
    yield ...
```

As I always try to advocate the YAGNI rule, this argument is not a valid point; you already implemented a signal trap but failed to name it properly. That’s it. Let’s refactor it and we are done with this merge request.